perses/rjmc/atom_mapping.py

```diff
@@ -0,0 +1,1475 @@
+
+        if len(atom_mappings) == 0:
+            _logger.warning('No maps found. Try relaxing match criteria or setting allow_ring_breaking to True')
+            return None
+
+        # Render set of AtomMapping to a list to return
+        return list(atom_mappings)
+
+    def get_best_mapping(self, old_mol, new_mol):
+        """Retrieve the best mapping between old and new molecules.
+
+        Parameters
+        ----------
+        old_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The initial molecule for the transformation.
+        new_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The final molecule for the transformation.
+
+        Returns
+        -------
+        atom_mapping : AtomMapping
+            Atom mapping with the best score
+
+        Examples
+        --------
+        Retrieve best-scoring mapping between ethane and ethanol
+
+        >>> from openff.toolkit.topology import Molecule
+        >>> ethane = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])([H:7])')
+        >>> ethanol = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])[O:7][H:8]')
+        >>> atom_mapping = atom_mapper.get_best_mapping(ethane, ethanol)
+
+        """
+        import time
+        initial_time = time.time()
+
+        import numpy as np
+        atom_mappings = self.get_all_mappings(old_mol, new_mol)
+        if (atom_mappings is None) or len(atom_mappings)==0:
+            return None
+
+        scores = np.array([ self.score_mapping(atom_mapping) for atom_mapping in atom_mappings ])
+        best_map_index = np.argmax(scores)
+
+        elapsed_time = time.time() - initial_time
+        _logger.debug(f'get_best_mapping took {elapsed_time:.3f} s')
+
+        return atom_mappings[best_map_index]
+
+    def get_sampled_mapping(self, old_mol, new_mol):
+        """Stochastically generate a mapping between old and new molecules selected proportional to its score.
+
+        Parameters
+        ----------
+        old_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The initial molecule for the transformation.
+        new_mol : openff.toolkit.topology.Molecule or openeye.oechem.OEMol
+            The final molecule for the transformation.
+
+        Returns
+        -------
+        atom_mapping : AtomMapping
+            Atom mapping with the best score
+
+        Examples
+        --------
+        Sample a mapping stochasticaly between ethane and ethanol
+
+        >>> from openff.toolkit.topology import Molecule
+        >>> ethane = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])([H:7])')
+        >>> ethanol = Molecule.from_smiles('[C:0]([H:1])([H:2])([H:3])[C:4]([H:5])([H:6])[O:7][H:8]')
+        >>> atom_mapping = atom_mapper.get_sampled_mapping(ethane, ethanol)
+
+        """
+        import numpy as np
+        atom_mappings = self.get_all_maps(old_mol, new_mol)
```

Comment:
I believe `get_all_maps` doesn't exist as a public/user-visible method. Is this supposed to be `_get_all_maps`? Now that makes me wonder if this method is being tested at all (probably not).

Reply:
Actually I now realize as well that the `_get_all_maps` static method has an undefined reference to `self` if no atom or bond expressions are passed.

Reply:
I am hoping to refactor the private `_get_all_map()` method in a future PR---once I fully understand what it does!
This was just a partial refactor, cleaning up the public API, eliminating references to the private API when possible, and refactoring the bits of the code I could understand. We'll need @dominicrufa's help (after he returns from his internship) for the next phase of refactoring/cleanup.